src/gp_engine.cpp

```diff
@@ -111,12 +111,13 @@ void GPEngine::operator()(const GPOperations::UpdateSBNProbabilities& op) {
     return;
   }
   // else
-  EigenVectorXd log_likelihoods = GetPerGPCSPLogLikelihoods();
-  Eigen::VectorBlock<EigenVectorXd> segment =
-      log_likelihoods.segment(op.start_, range_length);
-  const double log_norm = NumericalUtils::LogSum(segment);
-  segment.array() -= log_norm;
-  q_.segment(op.start_, range_length) = segment.array().exp();
+  EigenVectorXd log_likelihood =
+      GetPerGPCSPLogLikelihoods().segment(op.start_, range_length);
```

Comment:
This seems dangerous. `GetPerGPCSPLogLikelihoods` isn't an accessor-- it does calculation. Then we are returning a segment of that return value. If the full return value disappears then I think we'll have a dangling reference?

[This](https://stackoverflow.com/questions/61839333/c-dangling-reference-to-eigenmatrix-created-in-function) might be relevant.

Reply:
An alternative would be to write an override of `GetPerGPCSPLogLikelihoods` which takes a start and length. Then we'd just be doing the needed calculation.

Reply:
This is very interesting. Thanks for pointing this out. 

I agree with passing in start and length to limit the calculation only on the values that matter -- so I implemented an override for that function as you suggested.

Reply:
...but it doesn't look like you used it?